# 4.4 Task Data Wrangling & Subsetting

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create path

path = r'C:\Users\shahe\Desktop\Instacart Basket Analysis'

In [3]:
# Import orders data set

df_ords = pd.read_csv(r'C:\Users\shahe\Desktop\Instacart Basket Analysis\Data\Original data\orders.csv', index_col = False)

In [4]:
# Import products data set

df_prods = pd.read_csv(r'C:\Users\shahe\Desktop\Instacart Basket Analysis\Data\Original data\products.csv', index_col = False)

In [5]:
# Import departments data set

df_dep = pd.read_csv(r'C:\Users\shahe\Desktop\Instacart Basket Analysis\Data\Original data\departments.csv', index_col = False)

# Answer 2

In [6]:
# Another identifier variable in the df_ords dataframe that doesn’t need to be included in analysis as a numeric variable

df_ords['order_id'] = df_ords['order_id'].astype('str')

In [7]:
df_ords['user_id'].value_counts()

152340    100
185641    100
185524    100
81678     100
70922     100
         ... 
13046       4
70208       4
111440      4
31762       4
47980       4
Name: user_id, Length: 206209, dtype: int64

In [8]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [9]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(3)
memory usage: 182.7+ MB


# Answer 3

In [10]:
# Rename unintuitive name

df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [11]:
df_ords.head()

order_id user_id eval_set  order_number  orders_day_of_week  \
0  2539329       1    prior             1                   2   
1  2398795       1    prior             2                   3   
2   473747       1    prior             3                   3   
3  2254736       1    prior             4                   4   
4   431534       1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# Answer 4

In [12]:
# The most busiest hour for placing an orders

df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

Findings: peak time of placing order is 10:00 o'clock with 288,418 orders

# Answer 5

In [13]:
# Identifying department data set 

df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [14]:
df_dep_t = df_dep.T

In [15]:
# Transposing department data set

df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [16]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [17]:
# Take the first row of df_dep_t for the header

new_header = df_dep_t.iloc[0]

In [18]:
new_header

0    department
Name: department_id, dtype: object

In [19]:
df_dep_t_new = df_dep_t[1:]

In [20]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [21]:
# Take the data under the header row for a new df

df_dep_t_new = df_dep_t[1:]

In [22]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [23]:
df_dep_t_new.columns = new_header

In [24]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [25]:
data_dict = df_dep_t_new.to_dict('index')

In [26]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [27]:
# Determine the meaning behind a value of 4 in the "department_id" column within the df_prods data frame using a data dictionary.

print(data_dict.get('4'))

{'department': 'produce'}


# Answer 6

In [28]:
# Identifying breakfast item

print(data_dict.get('14'))

{'department': 'breakfast'}


In [29]:
df_14=df_prods[df_prods['department_id']==14]

In [30]:
df_list=df_14['product_name']

In [31]:
df_list

27                                       Wheat Chex Cereal
33                                                     NaN
67                                 Pancake Mix, Buttermilk
89                                            Smorz Cereal
210       Gluten Free Organic Cereal Coconut Maple Vanilla
                               ...                        
49330                              Cereal Variety Fun Pack
49395              Light and Fluffy Buttermilk Pancake Mix
49547                            Chocolate Cheerios Cereal
49637                 Shake 'N Pour Buttermilk Pancake Mix
49667    Ultra Protein Power Crunch Peanut Butter N' Ho...
Name: product_name, Length: 1116, dtype: object

In [32]:
# Listing all breakfast items

df_list.to_list()

['Wheat Chex Cereal',
 nan,
 'Pancake Mix, Buttermilk',
 'Smorz Cereal',
 'Gluten Free Organic Cereal Coconut Maple Vanilla',
 'Gluten Free Oatmeal, Variety Pack',
 'Honey Nut Oatmeal Squares',
 'Alpha-Bits Multigrain Cereal',
 'Waffles, Red Berry',
 'Original Sprouted Grains Protein & Fiber Hot Oatmeal',
 'Eco-Pac Kamut Puffs Cereal',
 'Cheerios Cereal',
 'Special K Vanilla Cappuccino Protein Breakfast Shakes',
 'Whole Wheat Fig Bar',
 'Protein Blend Honey Oats & Seeds Cereal',
 'Granola, Veritas, Organic',
 'Light and Fluffy  Blueberry Pancake Mix',
 'Special K Original Cereal',
 'Multigrain Peanut Butter Cereal',
 'Sweet Lemon Rain Dance Bar',
 'Organic Promise Strawberry Fields Cereal',
 'Quick Cooking Steel Cut Oats',
 'Complete Funfetti Buttermilk Pancake & Waffle Mix',
 'Healthy Grains Oats & Honey Clusters with Toasted Coconut Gluten Free',
 'Crunchy Granola Bars Variety Pack',
 'Nutri-Grain Harvest Hearty Breakfast Bars Country Strawberry',
 'Mallow Oats Cereals',
 'Shake n Po

In [33]:
df_list.to_csv("breakfast.csv")

# Answer 7

In [34]:
# Find all observations from the entire data frame that include items from the following departments: 
# alcohol, deli, beverages, and meat/seafood

df_items=data_dict.get('5') | data_dict.get('7') | data_dict.get('12') | data_dict.get('20')

In [35]:
df_items=df_prods[(df_prods['department_id']==5) | (df_prods['department_id']==7) | (df_prods['department_id']==12) | (df_prods['department_id']==20)]

In [36]:
df_items.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

In [37]:
df_items['product_name'].to_list()

['Robust Golden Unsweetened Oolong Tea',
 'Pure Coconut Water With Orange',
 'Sparkling Orange Juice & Prickly Pear Beverage',
 'Peach Mango Juice',
 'Rendered Duck Fat',
 'Pomegranate Cranberry & Aloe Vera Enrich Drink',
 'Organic Turkey Burgers',
 'Italian Herb Porcini Mushrooms Chicken Sausage',
 'Daily Tangerine Citrus Flavored Beverage',
 'Beef Hot Links Beef Smoked Sausage With Chile Peppers',
 'Sparkling Raspberry Seltzer',
 'Vegetarian Grain Meat Sausages Italian - 4 CT',
 'Mirabelle Brut Rose',
 'Whole Leaf Pure Aloe With Lemon Juice',
 'Premium Deli Oven Roasted Turkey Breast',
 nan,
 'Jasmine Tea Unfiltered Ginger Ale',
 "Artisan Chick'n & Apple Sausage",
 'All Natural 100% Apple Juice',
 'Lamb Shank',
 'Soppressata Piccante',
 'Classics Earl Grey Tea',
 'Organic Chamomile Lemon Tea',
 'Grape Leaf Hummus Wrap',
 'Chardonnay Paso Robles',
 'Olli Napoli Salami',
 'Organic Ranch Veggie Dip',
 'Brut Rosé',
 'Smoked Ham & Oven Roasted Turkey Sub Sandwich Kit',
 'Sparkling Blush G

# Answer 8

In [38]:
# Total counts in your data frames

df_items.shape

(7650, 5)

# Answer 9

In [39]:
# Extract all the information you can about the user 1

user_1 = df_ords[df_ords['user_id']=='1']

In [40]:
user_1

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   
5   3367565       1    prior             6                   2   
6    550135       1    prior             7                   1   
7   3108588       1    prior             8                   1   
8   2295261       1    prior             9                   1   
9   2550362       1    prior            10                   4   
10  1187899       1    train            11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

# Answer 10

In [41]:
# Some details and basic stats about this user’s behavior

user_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

# Answer 12

In [42]:
# Export wrangled orders data set

df_ords.to_csv(os.path.join(path,'Data','Prepared data',"orders_wrangled.csv"))

# Answer 13

In [43]:
# Export wrangled department data set

df_dep_t_new.to_csv(os.path.join(path,'Data','Prepared data',"department_wrangled.csv"))